opendevin/controller/agent_controller.py

```diff
@@ -248,8 +248,16 @@ def get_agent_state(self):
 
     async def start_delegate(self, action: AgentDelegateAction):
         agent_cls: Type[Agent] = Agent.get_cls(action.agent)
-        llm_config = config.get_llm_config_from_agent(action.agent)
-        llm = LLM(llm_config=llm_config)
+        if action.llm is not None:
+            # share the same llm instance as parent
+            # so the cost is accumulated together
+            llm = action.llm
```

Comment:
Assigning `llm` attribute to an action is a hack. I do see your point that metrics are not globally shared, which we should address by fixing somewhere else, for example,

```python
    async def update_state_after_step(self):
        # update metrics especially for cost
        self.state.metrics = self.agent.llm.metrics
```

Plus, this will break the logic that allows delegates to use a different LLM than its parent.

Reply:
I do think metric logging is less of a concern. This is more of a concern in the context of this PR: in evaluation, we typically fix the `llm` used for the whole run - Right now, without this hack, `BrowsingAgent` will, by default, try to use `[llm]` and its corresponding parameter (e.g., temperature = 1), which is not desired.

I'd figure it will be a much larger hack if we simply assign all child agent's LLM to the parent one - Any idea how we can better support this use case?

Reply:
I think we can set Agent-specific config, but as the delegation happens to more and more agents, adding one config for each agent will likely be a bit cumbersome..

Reply:
> BrowsingAgent will, by default, try to use [llm] and its corresponding parameter (e.g., temperature = 1), which is not desired.

We should set that in `config.toml` under `agent.BrowsingAgent` section, e.g.

```yaml
[agent.BrowsingAgent]
llm_config=eval_gpt_4o
```

Reply:
> but as the delegation happens to more and more agents, adding one config for each agent will likely be a bit cumbersome..

Then we should always do

```yaml
[llm.eval]
# change this every time we want to benchmark a different model
model=gpt_4o
temperature=0

[agent.BrowsingAgent]
llm_config=eval
```

I believe passing a config group via `--llm-config` CLI arg should be eventually deprecated. We should always use `config.toml`. C.C. @enyst what do you think?

A natural question to ask is, what if I want to run different benchmarks in parallel, each with a different LLM setting? The way to go IMO should be, we add a `--config` CLI arg that allows you to specify the path of config file. If absent, we use `config.toml`.

Reply:
I have another idea: we could reinterpret `--llm-config` CLI arg and let it override the default `config.llm` settings. Then you could let all agents (including delegates) use whatever is specified by `--llm-config`.

Reply:
Yeah i think using `--llm-config` to override everything sounds pretty reasonable to me! I'll update this PR to do this

Reply:
I just did! #2957

Reply:
Thanks and approved!

Reply:
Created #2958 to track the bug fix for metrics (cost).